File: sdk/python/test_cross_sdk_compat.py
"""K4-02: Cross-SDK compatibility test.

Validates Python SDK bincode encoding matches Rust golden vectors exactly.
The authoritative hex values come from core/src/transaction.rs
test_cross_sdk_message_golden_vector and test_cross_sdk_transaction_golden_vector.
"""

import sys
import os
import hashlib

# Add parent directory so we can import lichen package
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from lichen.publickey import PublicKey
from lichen.bincode import encode_message, encode_transaction, EncodedInstruction
from lichen.pq import PqPublicKey, PqSignature
from lichen.transaction import TransactionBuilder


# --- Deterministic test data (same as Rust golden vector tests) ---
program_id = PublicKey(bytes([0x01] * 32))
account0   = PublicKey(bytes([0x02] * 32))
data       = bytes([0x00, 0x01, 0x02, 0x03])
blockhash  = "aa" * 32   # 32 bytes as hex string
signature  = PqSignature(
    scheme_version=0x01,
    public_key=PqPublicKey(0x01, bytes([0xBB]) * 1952),
    sig=bytes([0xBB]) * 3309,
)

ix = EncodedInstruction(program_id=program_id, accounts=[account0], data=data)


def test_message_golden_vector():
    """Message encoding must match Rust bincode output exactly."""
    msg_bytes = encode_message([ix], blockhash)
    got = msg_bytes.hex()

    # Authoritative value from Rust test_cross_sdk_message_golden_vector
    expected = (
        "0100000000000000"                                            # Vec<Ix> len = 1
        "0101010101010101010101010101010101010101010101010101010101010101"  # program_id
        "0100000000000000"                                            # Vec<Pubkey> len = 1
        "0202020202020202020202020202020202020202020202020202020202020202"  # accounts[0]
        "040000000000000000010203"                                    # Vec<u8> len=4 + data
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"  # blockhash
        "0000"                                                        # compute_budget: None + compute_unit_price: None
    )

    assert got == expected, (
        f"K4-02 PYTHON MESSAGE GOLDEN VECTOR MISMATCH!\n"
        f"Got:      {got}\n"
        f"Expected: {expected}"
    )
    print("  ✓ Message golden vector matches Rust")


def test_transaction_golden_vector():
    """Transaction encoding must match Rust bincode output length and hash."""
    msg_bytes = encode_message([ix], blockhash)
    tx_bytes = encode_transaction([signature], msg_bytes)
    tx_hash = hashlib.sha256(tx_bytes).hexdigest()

    assert len(tx_bytes) == 5417, (
        f"K4-02 PYTHON TX LENGTH MISMATCH!\n"
        f"Got:      {len(tx_bytes)}\n"
        f"Expected: 5417"
    )
    assert tx_hash == "9d0eec7b657276b828c265995ce78b41a3e19b17ab354b11f37254bbc4ee2a91", (
        f"K4-02 PYTHON TX HASH MISMATCH!\n"
        f"Got:      {tx_hash}\n"
        f"Expected: 9d0eec7b657276b828c265995ce78b41a3e19b17ab354b11f37254bbc4ee2a91"
    )
    print("  ✓ Transaction golden vector matches Rust length + hash")


def test_message_optional_compute_fields():
    """Optional compute fields must encode as bincode Option<u64>."""
    msg_bytes = encode_message([ix], blockhash, compute_budget=500_000, compute_unit_price=250)
    got = msg_bytes.hex()
    expected_suffix = (
        "01"                                                        # compute_budget: Some
        "20a1070000000000"                                          # 500_000 LE
        "01"                                                        # compute_unit_price: Some
        "fa00000000000000"                                          # 250 LE
    )

    assert got.endswith(expected_suffix), (
        f"K4-02 PYTHON COMPUTE OPTION MISMATCH!\n"
        f"Got suffix:      {got[-len(expected_suffix):]}\n"
        f"Expected suffix: {expected_suffix}"
    )
    print("  ✓ Optional compute fields encode correctly")


def test_transaction_builder_sets_compute_fields():
    """TransactionBuilder should preserve explicit compute settings on the message."""
    message = (
        TransactionBuilder()
        .add(TransactionBuilder.transfer(account0, program_id, 1))
        .set_recent_blockhash(blockhash)
        .set_compute_budget(500_000)
        .set_compute_unit_price(250)
        .build()
    )

    assert message.compute_budget == 500_000
    assert message.compute_unit_price == 250
    print("  ✓ TransactionBuilder preserves compute settings")


if __name__ == "__main__":
    test_message_golden_vector()
    test_transaction_golden_vector()
    test_message_optional_compute_fields()
    test_transaction_builder_sets_compute_fields()
    print("K4-02: All Python cross-SDK compatibility tests passed")

File: sdk/js/test_cross_sdk_compat.js
// K4-02: Cross-SDK compatibility test
// Validates JS SDK bincode encoding matches Rust golden vectors exactly.
// The authoritative hex values come from core/src/transaction.rs
// test_cross_sdk_message_golden_vector and test_cross_sdk_transaction_golden_vector.

import assert from 'node:assert';
import crypto from 'node:crypto';

// --- Inline minimal encoder (mirrors bincode.ts without TS/ESM deps) ---

function encodeU64LE(value) {
  const out = new Uint8Array(8);
  const view = new DataView(out.buffer);
  view.setBigUint64(0, BigInt(value), true);
  return out;
}

function hexToBytes(hex) {
  const clean = hex.startsWith('0x') ? hex.slice(2) : hex;
  const out = new Uint8Array(clean.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

function bytesToHex(bytes) {
  return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
}

function concat(parts) {
  const total = parts.reduce((sum, p) => sum + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) { out.set(p, offset); offset += p.length; }
  return out;
}

function encodeVec(items) {
  return concat([encodeU64LE(items.length), ...items]);
}

function encodeBytes(data) {
  return concat([encodeU64LE(data.length), data]);
}

function encodePubkey(bytes32) {
  if (bytes32.length !== 32) throw new Error('Pubkey must be 32 bytes');
  return bytes32;
}

function encodeInstruction(ix) {
  const programId = encodePubkey(ix.programId);
  const accounts = encodeVec(ix.accounts.map(encodePubkey));
  const data = encodeBytes(ix.data);
  return concat([programId, accounts, data]);
}

function encodeMessage(msg) {
  const instructions = encodeVec(msg.instructions.map(encodeInstruction));
  const blockhash = msg.recentBlockhash;
  if (blockhash.length !== 32) throw new Error('Blockhash must be 32 bytes');
  // compute_budget: Option<u64> (None = 0x00)
  const cb = new Uint8Array([0x00]);
  // compute_unit_price: Option<u64> (None = 0x00)
  const cup = new Uint8Array([0x00]);
  return concat([instructions, blockhash, cb, cup]);
}

function encodeU32LE(value) {
  const out = new Uint8Array(4);
  const view = new DataView(out.buffer);
  view.setUint32(0, value, true);
  return out;
}

function encodeU8(value) {
  return Uint8Array.of(value & 0xff);
}

function encodePqPublicKey(publicKey) {
  return concat([encodeU8(publicKey.scheme_version), encodeBytes(publicKey.bytes)]);
}

function encodePqSignature(signature) {
  return concat([
    encodeU8(signature.scheme_version),
    encodePqPublicKey(signature.public_key),
    encodeBytes(signature.sig),
  ]);
}

function encodeTransaction(sigs, messageBytes) {
  const sigParts = sigs.map(encodePqSignature);
  // tx_type: Native = 0 (u32 LE)
  const txType = encodeU32LE(0);
  return concat([encodeU64LE(sigParts.length), ...sigParts, messageBytes, txType]);
}

// --- Deterministic test data (same as Rust golden vector tests) ---

const programId = new Uint8Array(32).fill(0x01);
const account0 = new Uint8Array(32).fill(0x02);
const data = new Uint8Array([0x00, 0x01, 0x02, 0x03]);
const blockhash = new Uint8Array(32).fill(0xAA);
const sig = {
  scheme_version: 0x01,
  public_key: {
    scheme_version: 0x01,
    bytes: new Uint8Array(1952).fill(0xBB),
  },
  sig: new Uint8Array(3309).fill(0xBB),
};

const message = {
  instructions: [{ programId, accounts: [account0], data }],
  recentBlockhash: blockhash,
};

// --- Test 1: Message golden vector ---
{
  const msgBytes = encodeMessage(message);
  const hex = bytesToHex(msgBytes);

  // Authoritative value from Rust test_cross_sdk_message_golden_vector
  const expected =
    '0100000000000000' +                                          // Vec<Ix> len = 1
    '0101010101010101010101010101010101010101010101010101010101010101' + // program_id
    '0100000000000000' +                                          // Vec<Pubkey> len = 1
    '0202020202020202020202020202020202020202020202020202020202020202' + // accounts[0]
    '040000000000000000010203' +                                  // Vec<u8> len=4 + data
    'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' + // blockhash
    '0000';                                                       // compute_budget: None + compute_unit_price: None

  assert.strictEqual(hex, expected,
    `K4-02 JS MESSAGE GOLDEN VECTOR MISMATCH!\nGot:      ${hex}\nExpected: ${expected}`);
  console.log('  ✓ Message golden vector matches Rust');
}

// --- Test 2: Transaction golden vector ---
{
  const msgBytes = encodeMessage(message);
  const txBytes = encodeTransaction([sig], msgBytes);
  const hex = bytesToHex(txBytes);

  const hash = crypto.createHash('sha256').update(Buffer.from(txBytes)).digest('hex');

  assert.strictEqual(txBytes.length, 5417,
    `K4-02 JS TX LENGTH MISMATCH!\nGot:      ${txBytes.length}\nExpected: 5417`);
  assert.strictEqual(hash, '9d0eec7b657276b828c265995ce78b41a3e19b17ab354b11f37254bbc4ee2a91',
    `K4-02 JS TX HASH MISMATCH!\nGot:      ${hash}\nExpected: 9d0eec7b657276b828c265995ce78b41a3e19b17ab354b11f37254bbc4ee2a91`);
  console.log('  ✓ Transaction golden vector matches Rust length + hash');
}

console.log('K4-02: All JS cross-SDK compatibility tests passed');
